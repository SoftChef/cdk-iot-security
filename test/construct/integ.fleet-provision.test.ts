import * as path from 'path';
import {
  App,
  Stack,
} from 'aws-cdk-lib';
import {
  Template,
} from 'aws-cdk-lib/assertions';
import {
  Runtime,
} from 'aws-cdk-lib/aws-lambda';
import {
  Bucket,
} from 'aws-cdk-lib/aws-s3';
import {
  FleetProvision,
} from '../../src';

const expectedResources: {
  [name: string]: string;
} = {
  provisoinRole: 'testcaseProvisionRoleFleetProvisioningRoletestcase678B85F8',
  fleetGeneratorRole: 'testcaseFleetGeneratortestcaseServiceRoleF1235FE0',
  fleetGeneratorPolicy: 'testcaseFleetGeneratortestcase5505B899',
  fleetGenerator: 'testcaseFleetGeneratortestcase12A380D3',
};

const expected: {
  [name: string]: string;
} = {
  lambdaFunctionRuntime: Runtime.NODEJS_14_X.toString(),
};

describe('Integration test', () => {

  test('FleetProvision', () => {
    process.env.BASE_PATH = __dirname;
    process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
    const app = new App();
    const stack = new Stack(app, 'test-stack');
    const name = 'test-case';
    const anotherStack = new Stack(app, 'another-stack');
    const bucket = new Bucket(anotherStack, 'userProvidedBucket');
    new FleetProvision(stack, name, {
      vault: {
        bucket: bucket,
        prefix: 'test',
      },
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
    template.resourceCountIs('AWS::Lambda::Function', 1);
    template.resourceCountIs('AWS::IAM::Policy', 2);
    template.resourceCountIs('AWS::IAM::Role', 2);
    template.hasResourceProperties('AWS::Lambda::Function', {
      Role: {
        'Fn::GetAtt': [
          expectedResources.fleetGeneratorRole,
          'Arn',
        ],
      },
      Environment: {
        Variables: {
          FLEET_PROVISIONING_ROLE_ARN: {
            'Fn::GetAtt': [
              expectedResources.provisoinRole,
              'Arn',
            ],
          },
        },
      },
      Runtime: expected.lambdaFunctionRuntime,
    });
    template.hasResourceProperties('AWS::IAM::Policy', {
      PolicyName: expectedResources.fleetGeneratorPolicy,
      PolicyDocument: {
        Statement: [
          {
            Action: [
              'iam:PassRole',
              'iam:CreatePolicy',
              'iam:CreateRole',
              'iam:TagRole',
              'iot:CreateProvisioningTemplate',
              'iot:CreatePolicy',
              'iot:CreateKeysAndCertificate',
              'iot:AttachPolicy',
              'iot:CreateRoleAlias',
            ],
            Effect: 'Allow',
            Resource: '*',
          },
        ],
        Version: '2012-10-17',
      },
      Roles: [
        {
          Ref: expectedResources.fleetGeneratorRole,
        },
      ],
    });
    template.hasResourceProperties('AWS::IAM::Role', {
      RoleName: 'ProvisionRoleName-FleetProvisioningRole-test-case',
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'iot.amazonaws.com',
            },
          },
        ],
        Version: '2012-10-17',
      },
      ManagedPolicyArns: [
        {
          'Fn::Join': [
            '',
            [
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSIoTThingsRegistration',
            ],
          ],
        },
      ],
    });
  });

  test('FleetProvision in Greengrass V2 mode', () => {
    process.env.BASE_PATH = __dirname;
    process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
    const app = new App();
    const stack = new Stack(app, 'test-stack');
    const name = 'test-case';
    const anotherStack = new Stack(app, 'another-stack');
    const bucket = new Bucket(anotherStack, 'userProvidedBucket');
    new FleetProvision(stack, name, {
      vault: {
        bucket: bucket,
        prefix: 'test',
      },
      enableGreengrassV2Mode: true,
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
    template.resourceCountIs('AWS::Lambda::Function', 1);
    template.resourceCountIs('AWS::IAM::Policy', 2);
    template.resourceCountIs('AWS::IAM::Role', 2);
    template.hasResourceProperties('AWS::Lambda::Function', {
      Role: {
        'Fn::GetAtt': [
          expectedResources.fleetGeneratorRole,
          'Arn',
        ],
      },
      Environment: {
        Variables: {
          FLEET_PROVISIONING_ROLE_ARN: {
            'Fn::GetAtt': [
              expectedResources.provisoinRole,
              'Arn',
            ],
          },
        },
      },
    });
    template.hasResourceProperties('AWS::IAM::Policy', {
      PolicyName: expectedResources.fleetGeneratorPolicy,
      PolicyDocument: {
        Statement: [
          {
            Action: [
              'iam:PassRole',
              'iam:CreatePolicy',
              'iam:CreateRole',
              'iam:TagRole',
              'iot:CreateProvisioningTemplate',
              'iot:CreatePolicy',
              'iot:CreateKeysAndCertificate',
              'iot:AttachPolicy',
              'iot:CreateRoleAlias',
            ],
            Effect: 'Allow',
            Resource: '*',
          },
        ],
        Version: '2012-10-17',
      },
      Roles: [
        {
          Ref: expectedResources.fleetGeneratorRole,
        },
      ],
    });
    template.hasResourceProperties('AWS::IAM::Role', {
      RoleName: 'ProvisionRoleName-FleetProvisioningRole-test-case',
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'iot.amazonaws.com',
            },
          },
        ],
        Version: '2012-10-17',
      },
      ManagedPolicyArns: [
        {
          'Fn::Join': [
            '',
            [
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSIoTThingsRegistration',
            ],
          ],
        },
      ],
    });
  });

});
