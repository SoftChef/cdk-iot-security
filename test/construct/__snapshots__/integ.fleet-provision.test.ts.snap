// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test FleetProvision 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbArtifactHashC47D4C30": Object {
      "Description": "Artifact hash for asset \\"b91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fb\\"",
      "Type": "String",
    },
    "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3BucketB63E8CFE": Object {
      "Description": "S3 bucket for asset \\"b91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fb\\"",
      "Type": "String",
    },
    "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3VersionKey60211803": Object {
      "Description": "S3 key for asset version \\"b91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fb\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testcaseFleetGeneratortestcase12A380D3": Object {
      "DependsOn": Array [
        "testcaseFleetGeneratortestcaseServiceRoleDefaultPolicy66DAC61D",
        "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3BucketB63E8CFE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3VersionKey60211803",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3VersionKey60211803",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": Object {
              "Fn::ImportValue": "another-stack:ExportsOutputRefuserProvidedBucket2349EF2D343936C6",
            },
            "BUCKET_PREFIX": "test",
            "FLEET_PROVISIONING_ROLE_ARN": Object {
              "Fn::GetAtt": Array [
                "testcaseProvisionRoleFleetProvisioningRoletestcase678B85F8",
                "Arn",
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testcaseFleetGeneratortestcase5505B899": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:PassRole",
                "iam:CreatePolicy",
                "iam:CreateRole",
                "iam:TagRole",
                "iot:CreateProvisioningTemplate",
                "iot:CreatePolicy",
                "iot:CreateKeysAndCertificate",
                "iot:AttachPolicy",
                "iot:CreateRoleAlias",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testcaseFleetGeneratortestcase5505B899",
        "Roles": Array [
          Object {
            "Ref": "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testcaseFleetGeneratortestcaseServiceRoleDefaultPolicy66DAC61D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttuserProvidedBucket2349EF2DArnA89CCA27",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttuserProvidedBucket2349EF2DArnA89CCA27",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testcaseFleetGeneratortestcaseServiceRoleDefaultPolicy66DAC61D",
        "Roles": Array [
          Object {
            "Ref": "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testcaseFleetGeneratortestcaseServiceRoleF1235FE0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testcaseProvisionRoleFleetProvisioningRoletestcase678B85F8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSIoTThingsRegistration",
              ],
            ],
          },
        ],
        "RoleName": "ProvisionRoleName-FleetProvisioningRole-test-case",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Integration test FleetProvision in Greengrass V2 mode 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbArtifactHashC47D4C30": Object {
      "Description": "Artifact hash for asset \\"b91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fb\\"",
      "Type": "String",
    },
    "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3BucketB63E8CFE": Object {
      "Description": "S3 bucket for asset \\"b91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fb\\"",
      "Type": "String",
    },
    "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3VersionKey60211803": Object {
      "Description": "S3 key for asset version \\"b91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fb\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testcaseFleetGeneratortestcase12A380D3": Object {
      "DependsOn": Array [
        "testcaseFleetGeneratortestcaseServiceRoleDefaultPolicy66DAC61D",
        "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3BucketB63E8CFE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3VersionKey60211803",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb91541c85e4343c88728be664ce3138442f1b4ddbf50a0f571ae84a46442b6fbS3VersionKey60211803",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": Object {
              "Fn::ImportValue": "another-stack:ExportsOutputRefuserProvidedBucket2349EF2D343936C6",
            },
            "BUCKET_PREFIX": "test",
            "ENABLE_GREENGRASS_V2_MODE": "true",
            "FLEET_PROVISIONING_ROLE_ARN": Object {
              "Fn::GetAtt": Array [
                "testcaseProvisionRoleFleetProvisioningRoletestcase678B85F8",
                "Arn",
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testcaseFleetGeneratortestcase5505B899": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:PassRole",
                "iam:CreatePolicy",
                "iam:CreateRole",
                "iam:TagRole",
                "iot:CreateProvisioningTemplate",
                "iot:CreatePolicy",
                "iot:CreateKeysAndCertificate",
                "iot:AttachPolicy",
                "iot:CreateRoleAlias",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testcaseFleetGeneratortestcase5505B899",
        "Roles": Array [
          Object {
            "Ref": "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testcaseFleetGeneratortestcaseServiceRoleDefaultPolicy66DAC61D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttuserProvidedBucket2349EF2DArnA89CCA27",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttuserProvidedBucket2349EF2DArnA89CCA27",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testcaseFleetGeneratortestcaseServiceRoleDefaultPolicy66DAC61D",
        "Roles": Array [
          Object {
            "Ref": "testcaseFleetGeneratortestcaseServiceRoleF1235FE0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testcaseFleetGeneratortestcaseServiceRoleF1235FE0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testcaseProvisionRoleFleetProvisioningRoletestcase678B85F8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSIoTThingsRegistration",
              ],
            ],
          },
        ],
        "RoleName": "ProvisionRoleName-FleetProvisioningRole-test-case",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
